packages/rpc-transport/src/json-rpc-transport-types.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * Public API.
+ */
+export type Transport<TApi> = TransportMethods<TApi>;
+export type ArmedTransport<Api, TResponse> = ArmedTransportMethods<Api, TResponse> & {
+    send(): Promise<TResponse>;
+};
+export type ArmedBatchTransport<Api, TResponses extends unknown[]> = ArmedBatchTransportMethods<Api, TResponses> & {
+    sendBatch(): Promise<TResponses>;
+};
+
+/**
+ * Private transport-building types.
+ */
+type TransportMethods<TApi> = {
+    [TMethodName in keyof TApi]: ArmedTransportReturner<TApi, ApiMethodImplementations<TApi, TMethodName>>;
+};
+type ArmedTransportMethods<TApi, TResponse> = {
+    [TMethodName in keyof TApi]: ArmedBatchTransportReturner<
+        TApi,
+        ApiMethodImplementations<TApi, TMethodName>,
+        [TResponse]
+    >;
+};
+type ArmedBatchTransportMethods<TApi, TResponses extends unknown[]> = {
+    [TMethodName in keyof TApi]: ArmedBatchTransportReturner<
+        TApi,
+        ApiMethodImplementations<TApi, TMethodName>,
+        TResponses
+    >;
+};
+type ApiMethodImplementations<TApi, TMethod extends keyof TApi> = Overloads<TApi[TMethod]>;
+type ArmedTransportReturner<TApi, TMethodImplementations> = UnionToIntersection<
```

Comment:
Is there something to the `ArmedTransport*` names that I'm missing, as in, other than their funniness?

Reply:
A regular transport hasn't accumulated any method calls yet. You can't call `send()` on it.

As soon as it has accumulated one, it's ‘armed.’ You can call `send()` on it.

Reply:
Ah, that makes sense, thanks!